sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java

```diff
@@ -0,0 +1,161 @@
+        ensureYellow();
+
+        execute("insert into t (x, y) values (?, ?)", new Object[]{1, "don't panic"});
+        refresh();
+
+        execute("select _seq_no, _primary_term from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        Long preUpdateSeqNo = (Long) response.rows()[0][0];
+        Long preUpdatePrimaryTerm = (Long) response.rows()[0][1];
+
+        execute("update t set y = ? where x = ? and _seq_no = ? and _primary_term = ?",
+                new Object[]{"ok now panic", 1, preUpdateSeqNo, preUpdatePrimaryTerm});
+        assertEquals(1L, response.rowCount());
+
+        // Validate that the row is really updated
+        refresh();
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("ok now panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithoutPrimaryKey() throws Exception {
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage(VersioninigValidationException.SEQ_NO_AND_PRIMARY_TERM_USAGE_MSG);
+        execute("create table t (x integer primary key, y string) with (number_of_replicas=0)");
+        ensureYellow();
+        execute("update t set y = ? where _seq_no = ? and _primary_term = ?",
+                new Object[]{"ok now panic", 1, 5});
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithConflict() throws Exception {
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
+
+        execute("insert into t (x, y) values (?, ?)", new Object[]{1, "don't panic"});
+        refresh();
+
+        execute("select _seq_no, _primary_term from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        Long preUpdateSeqNo = (Long) response.rows()[0][0];
+        Long preUpdatePrimaryTerm = (Long) response.rows()[0][1];
+
+        execute("update t set y = ? where x = ? and _seq_no = ? and _primary_term = ?",
+                new Object[]{"hopefully not updated", 1, preUpdateSeqNo + 4, preUpdatePrimaryTerm});
+        assertEquals(0L, response.rowCount());
+
+        // Validate that the row is really NOT updated
+        refresh();
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("don't panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testSelectWhereSeqNoAndPTWithoutPrimaryKey() throws Exception {
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage(VersioninigValidationException.SEQ_NO_AND_PRIMARY_TERM_USAGE_MSG);
+        execute("select _seq_no, _primary_term from t where y = 'hello' and _seq_no = 1 and _primary_term = 2");
+    }
+
+    @Test
+    public void testSelectGroupBySeqNoPrimaryTerm() throws Exception {
```

Comment:
What's special about this test case?

Reply:
I wondered that myself. I migrated some of the tests from https://github.com/crate/crate/blob/master/sql/src/test/java/io/crate/integrationtests/VersionHandlingIntegrationTest.java#L174 

I'm not sure what code path was meant to be covered by the grouping case to be honest. Happy to drop it if you think it's not worth it.

Reply:
yep I'd drop it. Maybe this is legacy from a time when our planner was less modular/compositional